Add tests for ?pttrf and ?pttrs

swallan opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Adds tests for #11223
#### What does this implement/fix?
<!--Please explain your changes.-->
This adds tests for the ?pttrf/s lapack routine
#### Additional information
<!--Any additional information you think is important.-->

swallan committed:
TST: Add tests for ?pttrf and ?pttrs
this commit adds two functions: one to test ?pttrf/s with randomly generated matricies of each dtype, and one to test the routines with the example problems from the NAG manual.

swallan committed:
TST: ?pttrf/s add error testing
Commit adds tests for mismatched array sizes, singular matricies, and non-spd matricies.

swallan commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,124 @@ def test_getc2_gesc2():
+        # diagonal d should always be real.
+        # for complex add 4 so it will be dominant
+        d = generate_random_dtype_array(n, DTYPES[1]) + 4
+    # diagonal e may be real or complex.
```
@mdhaber

Is there still an issue with this way of creating the d diagonal for complex types? It passes through pttrf but with info = 2, so I assume that there is still something wrong with the eigenvalues. Hopefully there's not something trivial that I overlooked.

Kai-Striega replied:
For the complex case, it could be due to the same reason as the NAG example fails. I'll confirm it tomorrow if that's the case.

swallan committed:
TST: fix typo and adhere to pep 8 style

swallan commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,124 @@ def test_getc2_gesc2():
+                         np.array([1+1j, 2-1j, 1-4j]),
+                         np.array([[64+16j,-16-32j],[93+62j, 61-66j], [78-80j,71-74j],[14-27j, 35+15j]]),
+                         np.array([[2-1j, -3-2j], [1+1j, 1+1j], [1-2j, 1-2j], [1-1j, 2+1j]])
+                         )])
```
@mdhaber

An area of concern that I have is about the complex nag test not working -

swallan commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,120 @@ def test_getc2_gesc2():
+                          np.array([1+1j, 2-1j, 1-4j]),
+                          np.array([[64+16j, -16-32j], [93+62j, 61-66j], [78-80j, 71-74j], [14-27j, 35+15j]]),
+                          np.array([[2-1j, -3-2j], [1+1j, 1+1j], [1-2j, 1-2j], [1-1j, 2+1j]])
+                         )])
```
@mdhaber

I'm a bit concerned about the complex type also not working with the provided NAG example. 

It fails immediately on the `assert_allclose(_d, d_expect,  rtol=rtol) `line (1768) when checking pttrf's output _d against the expected _d. I've checked over the numbers a few times, so I would hope that there's no typo messing this up.

Kai-Striega replied:
@swallan I think this one is on me. There is a case where the function parameters are a mix of real and complex variables. In this case the wrappers must be split into two distinct cases (see [gesvx](https://github.com/scipy/scipy/blob/b3a2071ce3952748c5e98ef7328d2b8eab77b2dc/scipy/linalg/flapack_gen.pyf.src#L102-L166)). I think this requires a similar split. I'll take an in-depth look at it tomorrow and get back to you.

Kai-Striega replied:
The above seems to be *an* issue. I've tried fixing it after which it seg faults. I'm going to need to spend some time to look into it, but I'll be flat out with work until the weekend.

swallan replied:
I’ll continue to investigate to find out what I can before then!

mdhaber replied:
@kai-striega by "the above" do you mean it's a wrapper issue, not a test issue?

Kai-Striega replied:
@swallan there were two mistakes (both on me)

    1. The argument `d` is real for both real and complex matrices. This requires using ``ctypereal`` or splitting the wrapper into two.
    2. The complex case has an additional ``uplo`` argument, which I didn't incude. This argument should also be tested.

One more thing; the expected value of x has an additional minus sign. I'll add it as a suggestion.

mdhaber replied:
I will share responsibility : ) I have only been looking at `d` documentation when working the issues. Good catch!

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,120 @@ def test_getc2_gesc2():
+def test_pttrf_pttrs_NAG(d, e, d_expect, e_expect, b, x_expect):
+    # test to assure that wrapper is consistent with NAG manual
+    # example problems: f07jdf and f07jef (real)
+    # hhttps://www.nag.com/numeric/fl/nagdoc_latest/html/f07/f07jdf.html
```
```suggestion
    # https://www.nag.com/numeric/fl/nagdoc_latest/html/f07/f07jdf.html
```

This isn't formatted as a url, as it has an extra `h`

swallan replied:
Fixed these issues, thanks!

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,120 @@ def test_getc2_gesc2():
         else:
             assert_array_almost_equal(desired_cplx.astype(dtype),
                                       x/scale, decimal=4)
+def generate_random_dtype_array(shape, dtype):
```
```suggestion


def generate_random_dtype_array(shape, dtype):
```
There should be two blank lines before top-level classes or functions. I've found a style checker becomes really useful for this kind of stuff. I found this using flake8 ``flake8 scipy/linalg/tests/test_linalg.py``.

swallan committed:
TST: Adhere to PEP8 and clarify comment

This commit fixes some PEP8 issues and a broken link typo.

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1743,20 +1745,24 @@ def test_pttrf_pttrs(dtype):
-                          np.array([[2-1j, -3-2j], [1+1j, 1+1j], [1-2j, 1-2j], [1-1j, 2+1j]])
+                          np.array([[64+16j, -16-32j], [93+62j, 61-66j],
+                                    [78-80j, 71-74j], [14-27j, 35+15j]]),
+                          np.array([[2-1j, -3-2j], [1+1j, 1+1j], [1-2j, 1-2j],
```
```suggestion
                          np.array([[2+1j, -3-2j], [1+1j, 1+1j], [1-2j, 1-2j],
```

swallan committed:
TST: pttrf/s NAG manual typo

Co-Authored-By: Kai Striega <kaistriega+github@gmail.com>

swallan committed:
Merge remote-tracking branch 'kai/lapack_support_pttrf_pttrs' into lapack_support_pttrf_pttrs

Kai-Striega commented:
@swallan are the tests now passing with the updated wrapper?

swallan commented:
Apologies, I should have followed up here. The NAG manual complex and real tests are both passing, but I have still run into issues with the complex randomly generated tests. Complex pttrf outputs don’t seem to recombine to original matrix A, and those factors then don’t make x with pttrs. So, I would assume it is either fully correct with some issue only regarding the complex randomly generated case, or there is another problem.

Kai-Striega commented:
Could you please copy and paste the values for which the complex test
fails? And could you try change the seed values, that way, if it is due to
the randomness, the test should pass.

On Sat., 15 Feb. 2020, 16:38 Sam Wallan, <notifications@github.com> wrote:

> Apologies, I should have followed up here. The NAG manual complex and real
> tests are both passing, but I have still run into issues with the complex
> randomly generated tests. Complex pttrf outputs don’t seem to recombine to
> original matrix A, and those factors then don’t make x with pttrs. So, I
> would assume it is either fully correct with some issue only regarding the
> complex randomly generated case, or there is another problem.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Kai-Striega/scipy/pull/3?email_source=notifications&email_token=AHFOJNNRUECJ4CDNNKIHOPLRC6SYRA5CNFSM4KLV4UCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEL3FF2I#issuecomment-586568425>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHFOJNNHPBK7XO4MDKNEJXLRC6SYRANCNFSM4KLV4UCA>
> .
>

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,126 @@ def test_getc2_gesc2():
+    else:
+        # diagonal d should always be real.
+        # for complex add 4 so it will be dominant
+        d = generate_random_dtype_array(n, DTYPES[1]) + 4
```
```suggestion
        d = (generate_random_dtype_array(n, np.float64) + 4).astype(dtype)
```

mdhaber replied:
Actually, although you only want real values, you should make sure `dtype` is respected. So for `complex64`, this should be of type `complex64` with all imaginary parts 0. This might be why the test is not passing?

swallan replied:
Interesting- right now it looks like the wrapper converts to real and raises a complex discard warning. I assume that passing in a complex type vs a real type array won't make any difference right now while the imaginary parts get stripped anyways. 

Changing this (with the current wrapper) doesn't appear to have an effect on the outputs, they still fail the tests.

mdhaber replied:
Let's take a look together this week?

swallan replied:
Sounds good! Office hours tomorrow are good for me. I can also come Wednesday morning.

mdhaber replied:
My comment above was wrong. We found the real issues (Hermitian = conjugate transpose) in office hours.

mdhaber commented:
I looked for issues with the complex randomly generated case and only found the one mentioned above. I checked that the eigenvalues are all positive (which was another thing that could have been wrong).

Kai-Striega commented:
> I looked for issues with the complex randomly generated case and only found the one mentioned above. I checked that the eigenvalues are all positive (which was another thing that could have been wrong).

Can you tell *why* it's wrong for that case? In terms of going forward, given that it's passing for other seed values, could we change the seed value? Otherwise, I will dig around the wrappers a bit to
see if there is any difference.

mdhaber commented:
I did not find a reason I am certain about, but I said it could be that the wrong data types was being used for `d`. You are welcome to check, or I can take a look with @swallan during office hours next week.

swallan commented:
@mdhaber I will see you tomorrow during office hours.

@Kai-Striega 

Here are the two arrays with seed 42: 
`d = array([4.37454012, 4.95071431, 4.73199394, 4.59865848, 4.15601864,
       4.15599452, 4.05808361, 4.86617615, 4.60111501, 4.70807258])` 

and 

`e = array([0.02058449+0.29122913j, 0.96990985+0.6118529j ,
       0.83244264+0.13949387j, 0.21233912+0.29214466j,
       0.18182497+0.36636186j, 0.1834045 +0.45606998j,
       0.30424225+0.785176j  , 0.52475643+0.19967379j,
       0.43194503+0.5142344j ], dtype=complex64)`

I also tried with a couple other seeds (52, 104,  and 553), and they all have similar issues, so I don't think it is an issue with that specific random case.

swallan committed:
TST: Tests pass, ensure A is hermetian

This commit has all the tests passing for each dtype.

swallan commented:
@Kai-Striega 

Tests for all 4 dtypes now pass, it was my mistake, not anything wrong with the wrapper!

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,123 @@ def test_getc2_gesc2():
+    # test that _x from pttrs matches the expected x
+    assert_allclose(x, _x, rtol=rtol, atol=atol)
+
+    # test that ValueError is raised with incompatible matrix shapes
+    with assert_raises(ValueError):
+        pttrf(d[:-1], e)
+    with assert_raises(ValueError):
+        pttrf(d, e[:-1])
```
One of the review requests I received on the previous PR was to separate independent tests into their own functions. I think that's a good idea in general. But didn't bring it up to stay consistent with the existing tests.

Given that we've changed it in the (now merged) branch, I think it would be good for us to have them split into distinct tests.

swallan replied:
Separated!

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,123 @@ def test_getc2_gesc2():
+                         )])
+def test_pttrf_pttrs_NAG(d, e, d_expect, e_expect, b, x_expect):
+    # test to assure that wrapper is consistent with NAG manual
+    # example problems: f07jdf and f07jef (real)
+    # https://www.nag.com/numeric/fl/nagdoc_latest/html/f07/f07jdf.html
+    # https://www.nag.com/numeric/fl/nagdoc_latest/html/f07/f07jef.html
+    # examples: f07jrf and f07csf (complex)
+    # https://www.nag.com/numeric/fl/nagdoc_latest/html/f07/f07jrf.html
+    # https://www.nag.com/numeric/fl/nagdoc_latest/html/f07/f07jsf.html
```
Good job documenting/referencing the NAG examples 👍

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,123 @@ def test_getc2_gesc2():
+
+
+@pytest.mark.parametrize("ddtype,dtype",
+                         list(zip(DTYPES, REAL_DTYPES + REAL_DTYPES)))
```
Here you don't need to construct a list. The `parameterize` will work with any iterable so simply using ``zip(...)`` will work without the extra overhead of creating a list.

mdhaber replied:
That's my fault. I didn't remember whether parametrize would accept any iterable.

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,123 @@ def test_getc2_gesc2():
+    _d, _e, info = pttrf(d, e)
+    np.testing.assert_(_d[info - 1] == 0,
+                       "?pttrf: _d[info-1] is {}, not the illegal value :0."
+                       .format(_d[info - 1]))
```
I'm not sure if @mdhaber has a strong opinion on this, but, I think it would be neater to user `assert_equal` or Python's standard `assert`. You won't need to worry if Python is built in debug mode because, unlike regular Python, Pytest always evaluates ``assert`` statements.

swallan replied:
Ok, I swapped assert_ out for those. Let me know if that's what you had in mind!

Kai-Striega replied:
They're looking good!

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,123 @@ def test_getc2_gesc2():
+    # test with non-spd matrix
+    d = generate_random_dtype_array((n,), dtype)
+    _d, _e, info = pttrf(d, e)
+    assert_(np.linalg.norm(d[info-1]) < 2 * np.linalg.norm(e[info-1]),
```
Same as above regarding assert

Kai-Striega commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,123 @@ def test_getc2_gesc2():
+    # https://www.nag.com/numeric/fl/nagdoc_latest/html/f07/f07jrf.html
+    # https://www.nag.com/numeric/fl/nagdoc_latest/html/f07/f07jsf.html
+    # NAG examples provide 4 decimals.
+    rtol = 1e-4
```
If you want to test to 4 decimal places this should be ``atol`` not ``rtol``.

Kai-Striega reviewed (commented):
@swallan I'm really happy with how this is coming together. I've left some minor review comments but overall it looks really good!

swallan committed:
TST: Break up tests to be more specific

Breaks up tests for pttrf pttrs into more specific functions. Uses assert_equal instead of assert_.

swallan committed:
TST: use assert instead of assert_()

Kai-Striega commented:
@mdhaber / @swallan are you happy that for me to merge the PR?

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,138 @@ def test_getc2_gesc2():
+    else:
+        # diagonal d should always be real.
+        # for complex add 4 so it will be dominant
+        d = (generate_random_dtype_array((n,), ddtype) + 4)
```
Why is this one in parentheses?
Let's add +4 regardless of whether it's complex or not to eliminate the branching.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,138 @@ def test_getc2_gesc2():
+    # diagonal e may be real or complex.
+    e = generate_random_dtype_array((n-1,), dtype)
+
+    # cast diagonals together into matrix
```
```suggestion
    # insert diagonals into matrix
```
arrange? assemble? nitpick, but I think there is a more appropriate word than cast.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,138 @@ def test_getc2_gesc2():
+    # test that the factors from pttrf can be recombined to make A
+    L = np.diag(_e, -1) + np.diag(np.ones(n))
+    D = np.diag(_d)
+
```
```suggestion
```

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,138 @@ def test_getc2_gesc2():
+    pttrf = get_lapack_funcs('pttrf', dtype=dtype)
+    d = generate_random_dtype_array((n,), ddtype) + 2
+    e = generate_random_dtype_array((n-1,), dtype)
+    # test that singular (row of all zeroes) matrix fails via info
```
This is singular because the diagonal element is zero. The element to the right of it is nonzero.

swallan replied:
Fixed in the next commit.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,138 @@ def test_getc2_gesc2():
+    d[0] = 0
+    _d, _e, info = pttrf(d, e)
+    assert_equal(_d[info - 1], 0,
+                 err_msg="?pttrf: _d[info-1] is {}, not the illegal value :0."
```
I'm confused by this error message.

swallan replied:
The message here was supposed to convey that the index that info returned does not contain the 0 that was put into d.

mdhaber commented on scipy/linalg/tests/test_lapack.py:
```diff
@@ -1654,3 +1654,138 @@ def test_getc2_gesc2():
+    d = generate_random_dtype_array((n,), ddtype)
+    _d, _e, info = pttrf(d, e)
+    assert np.linalg.norm(d[info-1]) < (2 * np.linalg.norm(e[info-1])), \
+        "idx {} of d should < e but are: {}, {} " \
```
I'm also confused by this error message.

swallan replied:
The idea for this came from this part of the documentation:

> if INFO = k, the leading minor of order k is not
               positive definite; if k < N, the factorization could not
               be completed, while if k = N, the factorization was
               completed, but D(N) <= 0.

So I checked that index to see that it is indeed not positive definite. Should I clarify / simplify the error message to say that that index should be positive definite?

mdhaber reviewed (approved):
@Kai-Striega I'm fine with you merging if you are willing to address the comments here; they're minor. The tests seem to do what they need to do.

swallan committed:
TST: Remove branching, clarify comments
